### PR TITLE
Code Guidelines

### PR DESCRIPTION
**Pull request by j-bjorne**
Initial commit of coding guidelines

### CODE REVIEW

File: coding_guidelines.md [hunk 1]
```diff
@@ -0,0 +1,500 @@
...
+}
+
+const isOpen = (hour24: number) => {
+  return hour24 > 6 && hour24 < 20;
+}
+
+const butcherShop = {
+  name: 'Meaty',
+  isOpen // AVOID!
+}
```
Comment (larsmoa) [inline coding_guidelines.md:83]: Why is this bad?

File: coding_guidelines.md [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+    this._owner = owner;
+  }
+
+  get name(): string {
+    return this._name;
+  }
+
+  get owner(): Owner {
+    return this._owner;
+  }
```
Comment (larsmoa) [inline coding_guidelines.md:151]: I think this is perfectly valid. This is how e.g. C# works too and should be familiar. If we want to be explicit about this sort of stuff, we should return `[Readonly<Owner>](https://www.typescriptlang.org/docs/handbook/utility-types.html#readonlyt)` instead.

File: coding_guidelines.md [hunk 3]
```diff
@@ -0,0 +1,500 @@
...
+
+owner.name = 'Billy';
+
+console.log(candyShop.name); // CandyPlusPlus
+console.log(candyShop.owner.name); // Billy
+```
+> In this example it might make more sense if the owner name was `readonly` and the owner had both a setter and a getter.
+
+### [Private Variables](#private-variables)
+Prefix private variables with `_`, e.g. `_name`, `_age`, `_height`.
```
Comment (larsmoa) [inline coding_guidelines.md:169]: 👍

File: coding_guidelines.md [hunk 4]
```diff
@@ -0,0 +1,500 @@
...
+
+### [Naming Guidelines](#naming-guidelines)
```
Comment (larsmoa) [inline coding_guidelines.md:171]: I don't think we can avoid bad code using a coding style tbh...

File: coding_guidelines.md [hunk 5]
```diff
@@ -0,0 +1,500 @@
...
+...
+```
+Now in addition to describing what happens inside the function it also matches with the return type of the function.
+
+# Arrays
+
+### [Array Instantiation](#array-instantiation)
+For array instantiation you should use the literal syntax instead of `new` constructor.
+```typescript
+// const items: number[] = new Array(); Don't do this.
```
Comment (larsmoa) [inline coding_guidelines.md:246]: Makes perfect sense to use `new Array(...)`, e.g. when you want to specify size up front

File: coding_guidelines.md [hunk 6]
```diff
@@ -0,0 +1,500 @@
...
+### [Expression Vs Declaration](#expression-vs-declaration)
+When defining a function we prefer function declaration over function expression, and when dealing with assigned functions we prefer using using an anonymous [arrow function](#arrow-functions) expression.
+```typescript
+function helloWorld(): void {} // Yes
+const helloWorld = function(): Void => {}; // No
+const helloWorld = (): void => {}; // No
+```
+> **Note:** Performance based on numbers from: https://jsperf.com/function-vs-arrow-function/14.
+
+> **Warning:** Declaring a function in a non-function block (if, while etc) has had a history of having different behavior across browsers. It is recommended to assign the function to a variable instead.
```
Comment (larsmoa) [inline coding_guidelines.md:269]: Doesn't webpack take care of this?

File: coding_guidelines.md [hunk 7]
```diff
@@ -0,0 +1,500 @@
...
+
+### [Default Parameters](#default-parameters)
+Instead of having mutating function arguments, consider using default parameters.
+```typescript
+// Worst implementation
+function handleThings(opts): void {
+  // No! We shouldn't mutate function arguments.
+  // Double bad: if opts is falsy it'll be set to an object which may
+  // be what you want but it can introduce subtle bugs.
+  opts = opts || {};
```
Comment (larsmoa) [inline coding_guidelines.md:282]: This is a pattern we use quite extensively... What do you propose we do with the usage we have? I think it's a quite nice pattern to use with options-objects, but I agree that falsy values is a bit risky. Maybe we should be explicit about matching undefined?

File: coding_guidelines.md [hunk 8]
```diff
@@ -0,0 +1,500 @@
...
+      this._targetTemperature = this._maximumTemperatureLimit;
+    }
+    else if(this._targetTemperature < this._minimumTemperaturLimit) {
+      this._targetTemperature = this._minimumTemperatureLimit;
+    }
+    return this._targetTemperature;
+  }
+
+  // Good function
+  public getTargetTemperatureAfterUpdatingTemperatureBy(degrees: number): number {
```
Comment (larsmoa) [inline coding_guidelines.md:363]: I disagree. What does 'update temperature by' mean? Increase? I think `increaseTemperatur(deltaDegrees: number): number` would be good...

However, I'm not sure if I agree that we should be this granular in our code style definitions.

File: coding_guidelines.md [hunk 9]
```diff
@@ -0,0 +1,500 @@
...
+}
+
+// Getting there!
+function authenticate(credentials: Credentials): boolean {
+  const { username, password } = credentials;
+  return this._apiService(username, password);
+}
+
+// And the reward goes to!
+function authenticate({ username, password }: Credentials): boolean {
```
Comment (larsmoa) [inline coding_guidelines.md:406]: Don't agree with this... The function declarations get awfully messy if they must be destructed in the argument list. I'd rather argument that if only a couple of fields of `Credentials` is used it's a sign of code smell and `username` and `password` should be passed as separate arguments.

File: coding_guidelines.md [hunk 10]
```diff
@@ -0,0 +1,500 @@
...
+# Functions
+
+### [Expression Vs Declaration](#expression-vs-declaration)
+When defining a function we prefer function declaration over function expression, and when dealing with assigned functions we prefer using using an anonymous [arrow function](#arrow-functions) expression.
+```typescript
+function helloWorld(): void {} // Yes
+const helloWorld = function(): Void => {}; // No
+const helloWorld = (): void => {}; // No
+```
+> **Note:** Performance based on numbers from: https://jsperf.com/function-vs-arrow-function/14.
```
Comment (larsmoa) [inline coding_guidelines.md:267]: I don't think that test is conclusive - my results vary with the choice of browser

File: coding_guidelines.md (hunk 1 above)
Comment (j-bjorne) [reply coding_guidelines.md]: This goes against the rule we set up that we want classes. Literal syntax is what we would use if we went for a more functional programming approach.

File: coding_guidelines.md (hunk 5 above)
Comment (j-bjorne) [reply coding_guidelines.md]: No, actually that is bad. Saying new Array(5); actually doesn't create an array with the size 5, it just creates an array with length 5. This can actually trick you.

First of all you can no longer use length to see if an array is empty.
I would completely agree if there was some merit into doing this, but the way Arrays are implemented in js seems to not strengthen this.

https://stackoverflow.com/questions/18947892/creating-range-in-javascript-strange-syntax

Basically, if you do it this way, you can't find out if an array is empty, since lenght != 0 this way.

File: coding_guidelines.md (hunk 7 above)
Comment (j-bjorne) [reply coding_guidelines.md:406]: Well, as proposed further down is to rather add the default to the function variable, and yes, that means if we agree on this we should change our old code to match the guideline.

File: coding_guidelines.md (hunk 8 above)
Comment (j-bjorne) [reply coding_guidelines.md]: I am saying, either a function is:
A) Super Granular
B) Documented with /** **/

File: coding_guidelines.md (hunk 9 above)
Comment (j-bjorne) [reply coding_guidelines.md:611]: At some point you need to destructure, whether you do it here or outside the function the fact of the matter is you either need:
```
const { username, password } = credentials
authenticate(username, password);

function authenticate(username: string, password: string) {}
```
or just
```
authenticate(credentials);

function authenticate({ username, password } : Credentials) {}
```

File: coding_guidelines.md (hunk 10 above)
Comment (j-bjorne) [reply coding_guidelines.md]: Sure, I just mean whatever codestyle we go for I would prefer we have 1, and not 2.

File: coding_guidelines.md (hunk 9 above)
Comment (larsmoa) [reply coding_guidelines.md:622]: or
```
authenticate(credentials);
function authenticate(credentials: Credentials) {
const { username, password } = credentials;
}
```
or just use `credentials.username` or `credentials.password` within the function. To me this isn't something we need to set in stone, but I dont think we should use destruction in function declarations.

Comment (j-bjorne) [reply coding_guidelines.md:622]: Destructuring and not destruction :D

Your example is covered as a do not, but we can change it to do. The first one:

```
function authenticate(credentials: Credentials): boolean {
  const username: string = credentials.username;
  const password: string = credentials.password;
  return this._apiService(username, password);
}
```

We can try and avoid?

Comment (j-bjorne) [reply coding_guidelines.md:622]: Or are we saying, don't use destructuring at all?

**Comment by christjt** [comment]
What are we going to do with this PR?

**Comment by maksnester** [comment]
> What are we going to do with this PR?

I think that it's great that we discussed all that stuff but I don't really see any practical use from it, to be honest. No one will look into that guide while coding. I think any code issues should be resolved either with linter rules or during code reviews. Having different opinions is fine, if both opinions very strong we can discuss altogether. That's my opinion on that. I don't mind if we merge that file, I just don't expect everybody will follow it (we don't even have consensus on its content – it's big enough).
